samples/algorithms/integer-factorization/QuantumVizCounter.cs

```diff
@@ -0,0 +1,200 @@
+            foreach (var qubit in qubitsTraceData)
+            {
+                operation.Targets.Add((QubitReference)qubit);
+            }
+            operation.IsAdjoint = variant == OperationFunctor.Adjoint || variant == OperationFunctor.ControlledAdjoint;
+            callStack.Push(operation);
+        }
+        
+        /// <summary>
+        /// When the operation ends, it's popped from the stack and added as
+        /// child to the next top-most operation on the stack.
+        /// </summary>
+        public void OnOperationEnd(object[] returnedQubitsTraceData)
+        {
+            var lastOperation = callStack.Pop();
+            Debug.Assert(callStack.Count != 0, "Operation call stack must never get empty");
+
+            // Add operation to parent and also update qubits
+            callStack.Peek().AddChild(lastOperation);
+        }
+
+        /// <summary>
+        /// Adds an outer interface qubit to the circuit data structure, whenever
+        /// a new qubit is allocated.
+        /// </summary>
+        public void OnAllocate(object[] qubitsTraceData)
+        {
+            foreach (var qubit in qubitsTraceData)
+            {
+                Circuit.Qubits.Add(new Qubit { Id = ((QubitReference)qubit).Id });
+            }
+        }
+
+        public void OnRelease(object[] qubitsTraceData) {}
+
+        public void OnBorrow(object[] qubitsTraceData, long newQubitsAllocated) {}
+
+        public void OnReturn(object[] qubitsTraceData, long newQubitsAllocated) {}
+
+        /// <summary>
+        /// Tracing data is captured in the custom qubit type using the qubit's id.
+        /// </summary>
+        public object? NewTracingData(long qubitId) => new QubitReference { Id = qubitId };
+
+        /// <summary>
+        /// This ensures that the listener traces qubit data.
+        /// </summary>
+        public bool NeedsTracingDataInQubits => true;
+        #endregion
+    }
+
+    public class Circuit {
+        [JsonProperty(PropertyName = "qubits")]
+        public HashSet<Qubit> Qubits { get; set; } = new HashSet<Qubit>();
+        [JsonProperty(PropertyName = "operations")]
+        public List<Operation> Operations { get; set; } = new List<Operation>();
+    }
+
+    public class Operation {
+        [JsonProperty(PropertyName = "gate")]
+        public string Gate { get; set; } = string.Empty;
+        [JsonProperty(PropertyName = "displayArgs")]
+        public string DisplayArgs { get => $"CNOT: {CNOTCount} Clifford: {CliffordCount} R: {RCount} T: {TCount}"; }
+        [JsonProperty(PropertyName = "children")]
+        public List<Operation> Children { get; set; } = new List<Operation>();
+        [JsonProperty(PropertyName = "targets")]
+        public HashSet<QubitReference> Targets { get; set; } = new HashSet<QubitReference>();
+        [JsonProperty(PropertyName = "isAdjoint")]
+        public bool IsAdjoint { get; set; } = false;
+
+        [JsonIgnore]
```

Comment:
Minor: I was initially confused as for why we're ignoring these fields in JSON generation if these are the fields we're looking for - I'd leave a comment here, something like "// We're not outputting the resource count fields in the JSON, since we include them in DisplayArgs field"